File: 5-Problem-A-Day/244. Shortest Word Distance II.java
class WordDistance {

    Map<String, List<Integer>> locations;

    public WordDistance(String[] words) {
        locations = new HashMap<String, List<Integer>>();
        for (int i = 0; i < words.length; i++) {
            List<Integer> loc = locations.getOrDefault(words[i], new ArrayList<Integer>());
            loc.add(i);
            locations.put(words[i], loc);
        }
    }

    public int shortest(String word1, String word2) {
        List<Integer> loc1 = locations.get(word1), loc2 = locations.get(word2);
        int l1 = 0, l2 = 0, minDiff = Integer.MAX_VALUE;
        while (l1 < loc1.size() && l2 < loc2.size()) {
            minDiff = Math.min(minDiff, Math.abs(loc1.get(l1) - loc2.get(l2)));
            if (loc1.get(l1) < loc2.get(l2))
                l1++;
            else
                l2++;
        }
        return minDiff;
    }
}

/**
 * Your WordDistance object will be instantiated and called as such: WordDistance obj = new
 * WordDistance(words); int param_1 = obj.shortest(word1,word2);
 */

File: 2021/244. Shortest Word Distance II.cpp
class WordDistance {
public:
    unordered_map<string, vector<int>> indexMap;
    WordDistance(vector<string>& words) {
        for (int i = 0; i < words.size(); i++) {
            indexMap[words[i]].push_back(i);
        }
    }
    
    int shortest(string word1, string word2) {
        vector<int>& index1 = indexMap[word1], &index2 = indexMap[word2];
        int l1 = 0, l2 = 0, res = INT_MAX;
        while (l1 < index1.size() && l2 < index2.size()) {
            res = min(res, abs(index1[l1] - index2[l2]));
            if (index1[l1] > index2[l2]) {
                l2++;
            } else {
                l1++;
            }
        }
        return res;
    }
};

/**
 * Your WordDistance object will be instantiated and called as such:
 * WordDistance* obj = new WordDistance(words);
 * int param_1 = obj->shortest(word1,word2);
 */